app/Http/Controllers/PreferenceController.php

```diff
@@ -17,19 +17,15 @@ public function index()
         ]);
     }
 
-    public function store(Request $request)
+    public function update(Request $request)
     {
-        $request->validate([
-            'resource-language' => 'required',
-            'locale' => 'required',
-            'operating-system' => 'required',
-        ]);
+        $validKeys = $request->only(Preferences::getValidKeys());
 
-        Preferences::set([
-            'resource-language' => $request->input('resource-language'),
-            'locale' => $request->input('locale'),
-            'operating-system' => $request->input('operating-system'),
-        ]);
+        $filtered = array_filter($validKeys, function($val){
```

Comment:
Would this work?

```php
Preferences::set($request->only(Preferences::getValidKeys())->filter());
```

This is relying on `filter()` to reject all null.. which i *think* it does?

Reply:
Wouldn't the only() have to be "collected" first and then converted back to an array? e.g.
```
Preferences::set(collect($request->only(Preferences::getValidKeys())->filter()->toArray());
```

I went with an explicit callback because I thought that there might potentially be valid preferences with values of 0 which would get filtered out without one.  Too much?

Reply:
@ktsubouchi I'm OK if we go for the simpler version right now and have to refactor later if there's a valid `0`.

I'm bummed, I thought `only` came back with a collection.

```php
        Preferences::set(collect($request->only(Preferences::getValidKeys()))->filter());
```

This works, but it's definitely not the most readable thing *ever*.

Maybe this?

```php
        Preferences::set(
            collect($request->only(Preferences::getValidKeys()))->filter()
        );
```

If Preferences::set doesn't allow for a collection, I'd love to modify it so it does.

Reply:
Yeah, if `Preferences::set()` can accept a collection, that'll be cleanest.  I'll check into that.

Reply:
Made a mistake on the last push...I'll fix that.

Reply:
Okay, done.  Is that what you had in mind for the `Preferences::set()` method?